server/jobmanager/serverv1/jobmanager_server_test.go

```diff
@@ -0,0 +1,349 @@
+/*
+Copyright 2021 Andy Dalton
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package serverv1_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/adalton/teleport-exercise/pkg/jobmanager"
+	"github.com/adalton/teleport-exercise/pkg/jobmanager/jobmanagertest"
+	"github.com/adalton/teleport-exercise/server/jobmanager/serverv1"
+	"github.com/adalton/teleport-exercise/server/jobmanager/serverv1/testserverv1"
+	"github.com/adalton/teleport-exercise/service/jobmanager/jobmanagerv1"
+	"github.com/adalton/teleport-exercise/util/grpcutil"
+	"github.com/google/uuid"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func Test_jobmanagerServer_Start_NoUserID(t *testing.T) {
+	jobManager := jobmanager.NewManagerDetailed(jobmanagertest.NewMockJob, nil)
+	server := serverv1.NewJobManagerServerDetailed(jobManager)
+
+	_, err := server.Start(context.Background(), &jobmanagerv1.JobCreationRequest{})
```

Comment:
These are unit tests, not integration tests, they test the behavior of one unit (the server code) in isolation from the rest of the system.  They are intentionally not integrated with the gRPC stack or with "real" job management.  They answer the question "does this small piece of code satisfy its contract?" in a way that's repeatable and fast.  If there's a failure in one of theses tests, it must be because of a failure in this one unit (or one of the mocks that it uses).

I have also provided integration tests that validate the client/server connectivity (see `test/server/server_test.go`).  Those integration tests together with these unit tests should provide decent functional coverage.  If you'd like, I can certainly extend the set of integration tests to cover all of the APIs, but I think the only additional value that that would add is a verification that the unit tests created the parameters to the APIs in the same way as the gPRC framework would have.

Reply:
I get the conceptual distinction, but I wouldn't recommend reaching for the server implementation directly in tests regardless - it tends to become brittle and leave important behavior out. If you want to run a "lighter" stack, there are alternatives like [bufconn.Listener](https://pkg.go.dev/google.golang.org/grpc@v1.43.0/test/bufconn) (although I would do a normal Dial in most cases). It's also sometimes possible to reuse the same server for all tests in a package, which amortizes the cost of setup.

We have other integration-like tests from you, so feel free to do as you prefer here.